otter/convergence/service.py

```diff
@@ -56,99 +62,352 @@ def all_met(server, current_lb_nodes):
+    # nodes 'A' and 'B', assuming only a boolean `dirty` flag:
+
+    # - A: policy executed: groupN.dirty = True
+    # -               B: converge group N (repeat until done)
+    # - A: policy executed: groupN.dirty = True
+    # -               B: groupN.dirty = False
+
+    # Here, a policy was executed on group N twice, and A tried to mark it
+    # dirty twice. The problem is that when the converger node finished
+    # converging, it then marked it clean *after* node A tried to mark it dirty
+    # a second time. It's a small window of time, but if it happens at just the
+    # right moment, after the final iteration of convergence and before the
+    # group is marked clean, then the changes desired by the second policy
+    # execution will not happen.
+
+    # So instead of just a boolean flag, we'll takie advantage of ZK node
+    # versioning. When we mark a group as dirty, we'll create a node for it
+    # if it doesn't exist, and if it does exist, we'll write to it with
+    # `set`. The content doesn't matter - the only thing that does matter
+    # is the version, which will be incremented on every `set`
+    # operation. On the converger side, when it searches for dirty groups
+    # to converge, it will remember the version of the node. When
+    # convergence completes, it will delete the node ONLY if the version
+    # hasn't changed, with a `delete(path, version)` call.
+
+    # The effect of this is that if any process updates the dirty flag
+    # after it's already been created, the node won't be deleted, so
+    # convergence will pick up that group again. We don't need to keep
+    # track of exactly how many times a group has been marked dirty
+    # (i.e. how many times a policy has been executed or config has
+    # changed), only if there are _any_ outstanding requests for
+    # convergence, since convergence always uses the most recent data.
+
+    path = CONVERGENCE_DIRTY_PATH.format(tenant_id=tenant_id,
+                                         group_id=group_id)
+    eff = Effect(CreateOrSet(path=path, content='dirty'))
+    return eff
+
+
+def delete_divergent_flag(log, tenant_id, group_id, version):
+    """
+    Delete the dirty flag, if its version hasn't changed. See comment in
+    :func:`mark_divergent` for more info.
 
+    :return: Effect of None.
+    """
+    path = CONVERGENCE_DIRTY_PATH.format(tenant_id=tenant_id,
+                                         group_id=group_id)
+    return Effect(DeleteNode(path=path, version=version)).on(
+        success=lambda r: log.msg('mark-clean-success'),
+        error=lambda e: log.err(exc_info_to_failure(e), 'mark-clean-failure'))
 
-class Converger(Service, object):
-    """Converger service"""
 
-    def __init__(self, reactor, kz_client, dispatcher):
-        self._reactor = reactor
-        self._kz_client = kz_client
+class ConvergenceStarter(Service, object):
+    """
+    A service that allows registering interest in convergence, but does not
```

Comment:
Nitpick: "A service that accepts reports that a group as diverged and needs converging" or some such?